### PR TITLE
ace: fix cache-uncache conversion macros for sparse

### PR DESCRIPTION
**Pull request by lyakh**
cache_to_uncache() and uncache_to_cache() must be inline functions with proper argument annotations. Fix ACE definitions by alanogy with cAVS.

### CODE REVIEW

**Comment by lgirdwood** [comment]
@lyakh looks like this redues teh MTL sparse error to Zephyr only
```
Reminder: to see ALL warnings you must as usual build _from scratch_
/zep_workspace/zephyr/soc/xtensa/intel_adsp/ace/power.c:190:17: warning: incorrect type in argument 1 (different address spaces)
/zep_workspace/zephyr/subsys/logging/backends/log_backend_adsp_mtrace.c:69:13: warning: incorrect type in argument 1 (different address spaces)
/zep_workspace/zephyr/subsys/logging/backends/log_backend_adsp_mtrace.c:73:9: warning: incorrect type in argument 1 (different address spaces)
/zep_workspace/zephyr/subsys/logging/backends/log_backend_adsp_mtrace.c:78:67: warning: incorrect type in argument 1 (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:374:20: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:438:20: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:529:24: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:530:24: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:542:18: warning: incorrect type in assignment (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:543:18: warning: incorrect type in assignment (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:640:24: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:641:24: warning: incorrect type in initializer (different address spaces)
/zep_workspace/zephyr/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:802:33: warning: incorrect type in argument 1 (different address spaces)
Error: Process completed with exit code 1.
```

**Comment by lyakh** [comment]
> @lyakh looks like this redues teh MTL sparse error to Zephyr only

@lgirdwood exactly. The fixes aren't perfect, particularly to the lib-manager. I cut the propagation of `__sparse_cache` attributes at some point, by forced type-casts, but maybe we rather should propagate them further. E.g. maybe we should adapt `sys_mm_drv_*()` functions to use cached addresses on Xtensa, but I don't really know whether those addresses really must be cached. Also `ext_lib_get()->desc` pointers really are cached, but labeling them as such was a bit difficult, so I left them as "generic" / uncached for now too. If desired, we can change that in a follow-up

File: src/platform/intel/ace/include/ace/lib/mailbox.h [hunk 1]
```diff
@@ -39,29 +39,29 @@
 static inline void mailbox_sw_reg_write(size_t offset, uint32_t src)
 {
 	volatile uint32_t *ptr;
-	volatile uint32_t __sparse_cache *ptr_c;
+	uint32_t __sparse_cache *ptr_c;
 
-	ptr_c = (volatile uint32_t __sparse_cache *)(MAILBOX_SW_REG_BASE + offset);
+	ptr_c = (uint32_t __sparse_cache *)(MAILBOX_SW_REG_BASE + offset);
```
Comment (kv2019i) [inline src/platform/intel/ace/include/ace/lib/mailbox.h:44]: This looks a bit weird to me (we are casting volatile back), but compiler seems to be happy, so I guess this is ok. You live and learn.

File: src/platform/intel/ace/include/ace/lib/memory.h [hunk 2]
```diff
@@ -40,14 +40,19 @@
 #define SRAM_ALIAS_OFFSET		0x60000000
```
Comment (kv2019i) [inline src/platform/intel/ace/include/ace/lib/memory.h:40]: typoe: alanogy in commit message

File: src/library_manager/lib_manager.c [hunk 3]
```diff
@@ -366,11 +379,10 @@ int lib_manager_register_module(struct sof_man_fw_desc *desc, int module_id)
 	return ret;
 }
 
-static void lib_manager_dma_buffer_update(struct lib_manager_dma_buf *buffer, uintptr_t addr,
+static void lib_manager_dma_buffer_update(struct lib_manager_dma_buf *buffer,
 					  uint32_t size)
 {
 	buffer->size = size;
-	buffer->addr = addr;
 	buffer->w_ptr = buffer->addr;
```
Comment (kv2019i) [inline src/library_manager/lib_manager.c:386]: @lyakh  Is this removal intentional? This seems like a bigger change than just adding annotations?

File: src/library_manager/lib_manager.c [hunk 4]
```diff
@@ -388,14 +400,12 @@ static int lib_manager_dma_buffer_init(struct lib_manager_dma_buf *buffer, uint3
...
+	dcache_writeback_region((__sparse_force void __sparse_cache *)buffer->addr, size);
 
-	/* initialise the DMA buffer */
+	tr_dbg(&lib_manager_tr, "lib_manager_dma_buffer_init(): %#lx, %#lx",
+	       buffer->addr, buffer->end_addr);
 
-	lib_manager_dma_buffer_update(buffer, buffer->addr, size);
-	tr_dbg(&lib_manager_tr, "lib_manager_dma_buffer_init(): %p, %p",
-	       (void *)buffer->addr, (void *)buffer->end_addr);
 	return 0;
```
Comment (kv2019i) [inline src/library_manager/lib_manager.c:409]: Ditto here.

File: src/library_manager/lib_manager.c [hunk 5]
```diff
@@ -509,98 +519,98 @@ static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext,
...
 
 	while (copied_bytes < dst_size) {
+		uint32_t bytes_to_copy;
+		int ret;
+
 		if ((dst_size - copied_bytes) > MAN_MAX_SIZE_V1_8)
 			bytes_to_copy = MAN_MAX_SIZE_V1_8;
 		else
-			bytes_to_copy = (dst_size - copied_bytes);
+			bytes_to_copy = dst_size - copied_bytes;
```
Comment (kv2019i) [inline src/library_manager/lib_manager.c:542]: Unrelated changes creeping in.

File: src/platform/intel/ace/include/ace/lib/mailbox.h (hunk 1 above)
Comment (lyakh) [reply src/platform/intel/ace/include/ace/lib/mailbox.h:44]: well, first of all you know "volatile is evil." But the point here is that that pointer is never used to actually access memory, so no need for `volatile` there at all

File: src/library_manager/lib_manager.c (hunk 3 above)
Comment (lyakh) [reply src/library_manager/lib_manager.c:386]: yes, intentional. Yeah, there are some fixes like this. This assignment is redundant, this function was always called with `(buf, buf->addr,...)` so it was an assignment to itself. That's why I removed that `addr` argument completely. The change seemed small enough to me to integrate it...

File: src/library_manager/lib_manager.c (hunk 5 above)
Comment (lyakh) [reply src/library_manager/lib_manager.c:542]: ok, I could split this into 2: a clean up and sparse stuff, would that help?

Comment (kv2019i) [reply src/library_manager/lib_manager.c:542]: @lyakh you can keep it, it's just a bit distracting to have these in bigger patches.

File: src/platform/intel/ace/include/ace/lib/mailbox.h (hunk 1 above)
Comment (marc-hb) [reply src/platform/intel/ace/include/ace/lib/mailbox.h:44]: > "volatile is evil."

I think "harmful" is more popular hence a better search keyword: https://www.bing.com/search?q=volatile+considered+harmful

File: src/library_manager/lib_manager.c (hunk 5 above)
Comment (marc-hb) [reply src/library_manager/lib_manager.c:542]: It's not just a reviewer distraction, it also means no validation will never test these separately.

**Review by marc-hb — COMMENTED** [review]
>  looks like this redues teh MTL sparse error to Zephyr only

You forgot the link to the corresponding logs: https://github.com/thesofproject/sof/actions/runs/3695828516/jobs/6258707347

File: src/library_manager/lib_manager.c (hunk 3 above)
Comment (kv2019i) [reply src/library_manager/lib_manager.c:386]: Ack, @jxstelter @pjdobrowolski @softwarecki can you check?

**Review by kv2019i — APPROVED** [review]
Ack, thanks @lyakh . I'd like some of the lib-manager authors to go through the changes as well.

File: src/library_manager/lib_manager.c [hunk 6]
```diff
@@ -509,98 +519,98 @@ static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext,
...
 
-	/* allocate new buffer */
-	local_add = rballoc_align(0, caps, size, addr_align);
+	/* allocate new buffer: cached alias */
+	local_add = (__sparse_force void __sparse_cache *)rballoc_align(0, caps, size, addr_align);
 #endif
 	if (!local_add) {
 		tr_err(&lib_manager_tr, "dma_buffer_init(): alloc failed");
-		return -ENOMEM;
+		return NULL;
```
Comment (softwarecki) [inline src/library_manager/lib_manager.c:577]: You changed the function declarations to return void, so why are you returning a value?

File: src/library_manager/lib_manager.c [hunk 7]
```diff
@@ -509,98 +519,98 @@ static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext,
...
 	}
 
-	*address = (uint32_t)local_add;
-	return 0;
+	return local_add;
```
Comment (softwarecki) [inline src/library_manager/lib_manager.c:580]: same as above

File: src/library_manager/lib_manager.c (hunk 6 above)
Comment (lyakh) [reply src/library_manager/lib_manager.c:577]: @softwarecki it's returning `void *`

Comment (softwarecki) [reply src/library_manager/lib_manager.c:577]: my mistake, sorry

File: src/platform/intel/ace/include/ace/lib/mailbox.h (hunk 1 above)
Comment (lyakh) [reply src/platform/intel/ace/include/ace/lib/mailbox.h:44]: > if I am sharing memory between devices (to a context where the compiler doesn't see that the memory will be accessed) then volatile is appropriate.

@paulstelian97 not sure which "devices" you mean. I think `volatile` is supposed to tell the compiler that contents of that memory can change or have effects outside of the scope of this code, so don't optimise access out.
So e.g.
```
*v = 1;
*v = 2;
```
would normally look redundant to the compiler and it might throw the first assignment away, but if the `v` pointer is declared as `volatile`, it won't. Similar for
```
X = *v;
Y = *v;
```
the compiler could reuse the value already read from `v` in the first line to effectively make the second like `Y = X;` but if `v` is a `volatile` pointer, it won't.
In this case that pointer points to memory, shared with the host, so the compiler has no way to know that the host can modify that memory. But to the compiler `ptr` and `ptr_c` are two different pointers, pointing to two different addresses - with some fixed offset. The compiler has no idea that it's "the same" memory. So whatever we do with one of the two pointers will not tell anything to the compiler that something also affects the other pointer. And since `ptr_c` is never used to access memory, since it is never dereferenced, the `volatile` qualifier for it has no effect.

**Review by softwarecki — APPROVED** [review]

File: src/library_manager/lib_manager.c (hunk 3 above)
Comment (jxstelter) [reply src/library_manager/lib_manager.c:386]: Looks good.

**Review by jxstelter — APPROVED** [review]
